### ISSUE TITLE
transformUrlDelegate is not called for .wasm and mem files

### ISSUE THREAD

**Issue by Kouty** [mil-tokyo/webdnn#841]
`transformUrlDelegate(...)` is not called when forcing, for example, webassembly backend.

That's because .wasm file is requested by the corresponding emscripten .js file. 

Emscripten provides `Module.locateFile(...)` function to map urls that will be loaded. Unfortunately transformUrlDelegate(...) is run on the main thread, whereas locateFile(...) is run on a webworker. This means that when using transformUrlDelegate() for a kernel_{backendname}.js file, it should be also used for the corresponding emscripten binary file (kernel_{backendname}.wasm, or  kernel_{backendname}.mem). Also the mapped url should be sent, somehow, to the worker before it starts emscripten code.

Problems: impossibility to map all files that WebDNN will be loading.
Workers cannot be cross domain, unless the source code of the worker is a blob. Providing a blob means that the corresponding emscripten call to the .wasm file will fail. Not being able to map the .wasm file url, means that **WebDNN cannot work for cross domain scripts** with emscripten based backends.

P.S. `var Module = {};` inside webassembly_header.js makes impossible to define Module.locateFile() by prepending some js code to the generated .js file. Moreover that `var Module = {};` is indeed unnecessary.

**Comment by milhidaka** [comment]
In my assumption, only weight file is located at different server, because it is much larger than other files.
I understood that current implementation transforms kernel_*.js in addition to weight_*.bin, which is incomplete.
Is there any good idea to implement transforming wasm url? Embedding target url to Module.locateFile statically seems not very good way.

**Comment by Kouty** [comment]
> In my assumption, only weight file is located at different server

Not necessarily just those files. For long term caching you may want to add hash inside file name, and use tools like webpack to generate the proper urls. Anyway I expect that the weight file is the most likely to be on a different server.

In my case it is even a different problem: I load emscripten js files (the kernel .js file) inside a blob, and use WebDNN `transformUrlDelegate` returning the blob url. WebDNN consequently will load the blob url inside a worker. If the worker requests a file using a relative url, the resulting url will be something like: blob:https://....wasm. So the emscripten .js file will fail to load the .wasm/.mem file.

>Is there any good idea to implement transforming wasm url? Embedding target url to Module.locateFile statically seems not very good way.

The problem is that `locateFile` as well as `transformUrlDelegate` need to be  synchronous. So you must know the url of the weight file before `locateFile` gets called.
What I proposed earlier is to use `transformUrlDelegate` for `kernel_backendName.js` files and the corresponding weight file. Then send the file url to the worker so that it can be used inside locateFile. So it is not statically added to the Module, but you have to know the .wasm or .mem file name in advance.

As a patch, what I did right now is:
```javascript
const mapJson = JSON.stringify(fileMap);
const fileData = `var Module = {
    locateFile: function(url) {
      const fileMap = ${mapJson};
      if(fileMap[url] !== undefined) {
        return fileMap[url];
      } else {
        console.warn('Cannot find url for resource ' + url);
      }
      return url;
    }
  };`+ kernelJsFileContent;
const blob = new Blob([fileData], { type: 'application/javascript' });
const blobUrl = URL.createObjectURL(blob);
```
where fileMap is something like:
```javascript
const BASE_PATH = './network/';
const fileMap = {
...
  'kernels_asmjs.js': require(BASE_PATH + 'kernels_asmjs.js'),
  'kernels_asmjs.js.mem': require(BASE_PATH + 'kernels_asmjs.js.mem'),
  'kernels_fallback.js': require(BASE_PATH + 'kernels_fallback.js'),
  'kernels_webassembly.js': require(BASE_PATH + 'kernels_webassembly.js'),
  'kernels_webassembly.wasm': require(BASE_PATH + 'kernels_webassembly.wasm'),
};
```
and in webpack.config.js I use file-loader to obtain the URL of those files.

In order to do that, I removed `var Module = {}` from webassembly_header.js.

**Comment by milhidaka** [comment]
Thanks for the idea.
I implemented the solution in #842. It replaces the URL placeholder in `kernels_*.js` with transformed absolute URL.
Please check the branch.

**Comment by Kouty** [comment]
Sorry for the delay in the answer, yesterday I took one day off.

I confirm that the fix in #842 (tranform-url-wasm branch) uses `transformUrlDelegate` to map also .bin/.wasm files and  works correctly with my code.

So for me the issue can be closed.

Just a note for the record: using Blob to load workers makes debugging a little bit more difficult. Also I don't know if it may have other drawbacks. That said, I think this solution is a good one.
